website: Add Sphinx documentation to website using GitHub Actions

erin-le opened the pull request:
This PR adds a workflow file that builds the Sphinx documentation and pushes it to the gem5 website. Currently, it is set to run once a week and build documentation for the stable version of gem5. This PR also includes a Python script that adds frontmatter and makes other alterations to the HTML files generated by Sphinx so they are compatible with the gem5 website.

This was tested using my own forks of the gem5 website and the gem5 repo, so some issues may come up when this workflow is run with the gem5/website repo.

The workflow file also won't work right now, as the current stable version of gem5 doesn't have the `docs` folder needed to build Sphinx. This won't be a problem once v 24.1 comes out, however.

erin-le committed:
website: Add Sphinx documentation to website using GitHub Actions

powerjg commented:
This is great! I'm testing to see how well copilot works here :)

powerjg commented on .github/workflows/sphinx-to-website.yaml:
```diff
@@ -0,0 +1,71 @@
+                    git config user.email "41898282+github-actions[bot]@users.noreply.github.com"
+                    git add .
+                    git commit -m "Update Sphinx documentation" -a || true
+                    git push
```
Is this the right wat to do it? I feel like there is a special action type to commit to the repo.

powerjg commented on add-sphinx-docs.py:
```diff
@@ -0,0 +1,87 @@
+    f.write("---\n")
+    f.write(f'title: "Sphinx Documentation"\n')
+    f.write("parent: sphinx-docs\n")
+    f.write(f"permalink: /documentation/general_docs/sphinx_docs/index.html\n")
```
Maybe call it `documentation/general_docs/stdlib_api`?

powerjg reviewed (commented):
A couple of small things, but I really like this solution!

erin-le committed:
website: make requested changes; add index page to documentation.yml

This commit makes the requested changes to use a dedicated GitHub
action to commit and push, and to use a different permalink.
This commit also adds in the link to the index page to
documentation.yml, which was previously missed and is necessary
to have the index page be linked to by the website's sidebar.

powerjg commented on .github/workflows/sphinx-to-website.yaml:
```diff
@@ -0,0 +1,68 @@
+                    path: gem5/build/ALL
+                    key: gem5.opt
+              - name: Build gem5
+                # if: steps.cache-gem5.outputs.cache-hit != 'true'
+                # The entire build/ALL folder is cached, so try commenting this
+                # out to see if gem5 will build relatively quickly
```
Are you still planning on testing this? I think it's OK if it takes a while to build.

erin-le replied:
I did test this at one point; it looks like the time to build is about the same whether gem5 is built from scratch or if gem5 is built from a cached build/ALL directory. I'll remove the cache steps since they're only needed for testing.

powerjg commented on .github/workflows/sphinx-to-website.yaml:
```diff
@@ -0,0 +1,68 @@
+              - name: Commit documentation changes to gem5/website repo
+                uses: EndBug/add-and-commit@v9
+                with:
+                    add: '.'
```
Should we just add `website/_pages/documentation/general_docs/sphinx_docs`?

powerjg commented on add-sphinx-docs.py:
```diff
@@ -0,0 +1,86 @@
+import os
```
Needs copyright

powerjg commented on add-sphinx-docs.py:
```diff
@@ -0,0 +1,86 @@
+            html = remove_searchbar(html)
+            
+            for line in html:
+                f.write(line)
```
Needs newline.

Also probably needs to have black run on it.

powerjg reviewed (commented):
A couple more small things.

erin-le committed:
website: make requested changes

This commit changes the path to files to be committed and removes
the caching step from sphinx-to-website.yaml. It also makes
formatting changes to add-sphinx-docs.py and adds a copyright
header.
